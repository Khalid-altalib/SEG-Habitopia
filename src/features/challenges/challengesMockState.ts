import { ChallengesState } from "./challengesSlice";

// Mocks the challenge state for testing purposes
const challengesMockState: ChallengesState = {
    challenges: [
        {
            id: '1',
            name: 'Fitness',
            description: 'Be active',
            active: true

        },
        {
            id: '1',
            name: 'Study',
            description: 'Study hard',
            active: false
        }
    ],
    fetchChallenges: {
        loading: false,
        error: ""
    },
    joinChallenge: {
        loading: false,
        error: ""
    }
};

export default challengesMockState;