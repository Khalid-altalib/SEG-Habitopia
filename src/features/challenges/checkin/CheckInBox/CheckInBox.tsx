// Import necessary modules and types
import { Image, ZStack, useColorModeValue } from "native-base";
import React from "react";
import { TouchableOpacity } from "react-native";
import Text from "@components/Text";
import { CheckInSnippetItem, TextType } from "types";
import {
  sendCheckIn,
  setCheckedInSnippetItemStatus,
} from "@features/chat/chatSlice";
import { useDispatch } from "@app/hooks";
import { Toast } from "react-native-toast-message/lib/src/Toast";
import CheckInTime from "../CheckInTime";

type Props = {
  checkInSnippetItem: CheckInSnippetItem; // The CheckInSnippetItem object to display
};

const CheckInBox = (props: Props) => {
  const { checkInSnippetItem } = props;
  const { challenge, endDate, chatId } = checkInSnippetItem;
  const dispatch = useDispatch();

  const handlePress = async () => {
    dispatch(setCheckedInSnippetItemStatus(chatId)); // Update the status of the CheckInSnippetItem to "checked in"
    await dispatch(sendCheckIn(chatId)); // Send a check-in message to the chat
    Toast.show({
      // Show a success message
      type: "success",
      text1: "Successfully checked in!",
    });
  };

  // Calculate the time difference between the endDate and the current time, in hours
  const timeDifference = new Date(endDate).getTime() - new Date().getTime();
  const timeDifferenceHours = Math.round(timeDifference / (1000 * 60 * 60));

  // Render the CheckInBox component
  return (
    <TouchableOpacity
      style={{ marginRight: 25 }}
      onPress={handlePress}
      testID={"checkInBox"}
    >
      <ZStack size="full" style={{ aspectRatio: 1 }}>
        <Image
          source={{ uri: "https://picsum.photos/2000" }}
          alt={challenge.name}
          size="full"
          rounded="lg"
          position="absolute"
          style={{
            borderColor: useColorModeValue("#00000011", "#ffffff11"),
            borderWidth: 2,
          }}
        />
        <Text
          type={TextType.Regular}
          color="white"
          style={{
            textShadowColor: "black",
            textShadowRadius: 4,
            padding: 12.25,
          }}
        >
          {challenge.name}
        </Text>
        <CheckInTime hoursLeft={timeDifferenceHours} />
      </ZStack>
    </TouchableOpacity>
  );
};

export default CheckInBox;
