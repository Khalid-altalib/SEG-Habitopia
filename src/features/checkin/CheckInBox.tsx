import { Image, ZStack, useColorModeValue } from "native-base";
import React, { useState } from "react";
import { TouchableOpacity } from "react-native";
import CheckInTime from "./CheckInTime";
import Text from "@components/Text";
import { ChatParams, CheckInSnippetItem, TextType } from "types";
import { useNavigation } from "@react-navigation/native";
import { NativeStackNavigationProp } from "@react-navigation/native-stack";
import {
  sendCheckIn,
  setCheckedInSnippetItemStatus,
} from "@features/chat/chatSlice";
import { useDispatch } from "@app/hooks";
import { Toast } from "react-native-toast-message/lib/src/Toast";

type Props = {
  checkInSnippetItem: CheckInSnippetItem;
};

const CheckInBox = (props: Props) => {
  const { checkInSnippetItem } = props;
  const { challenge, endDate, chatId } = checkInSnippetItem;

  const dispatch = useDispatch();

  const handlePress = async () => {
    dispatch(setCheckedInSnippetItemStatus(chatId));
    await dispatch(sendCheckIn(chatId));
    Toast.show({
      type: "success",
      text1: "Successfully checked in!",
    });
  };

  const timeDifference = new Date(endDate).getTime() - new Date().getTime();
  const timeDifferenceHours = Math.round(timeDifference / (1000 * 60 * 60));

  return (
    <TouchableOpacity style={{ marginRight: 25 }} onPress={handlePress}>
      <ZStack size="full" style={{ aspectRatio: 1 }}>
        <Image
          source={{ uri: "https://picsum.photos/2000" }}
          alt={challenge.name}
          size="full"
          rounded="lg"
          position="absolute"
          style={{
            borderColor: useColorModeValue("#00000011", "#ffffff11"),
            borderWidth: 2,
          }}
        />
        <Text
          type={TextType.Regular}
          color="white"
          style={{
            textShadowColor: "black",
            textShadowRadius: 4,
            padding: 12.25,
          }}
        >
          {challenge.name}
        </Text>
        <CheckInTime hoursLeft={timeDifferenceHours} />
      </ZStack>
    </TouchableOpacity>
  );
};

export default CheckInBox;
