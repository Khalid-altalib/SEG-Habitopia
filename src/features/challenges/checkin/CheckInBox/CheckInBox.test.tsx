import { render } from "@testing-library/react-native";
import configureStore from "redux-mock-store";
import thunk from "redux-thunk";
import TestingWrapperNavigation from "@app/testingWrapperWithNavigation";
import CheckInBox from "./CheckInBox";
import challengesMockState from "@features/challenges/challengesMockState";

describe("CheckInBox", () => {
  const mockState = {
    challenges: challengesMockState,
    checkInSnippetItem: {
      challenge: challengesMockState.challenges[0],
      endDate: "",
      chatId: "",
    },
  };
  const mockStore = configureStore([thunk])(mockState);
  const props = {
    checkInSnippetItem: mockState.checkInSnippetItem,
  };

  it("renders the check-in box correctly", async () => {
    const component = render(
      <TestingWrapperNavigation store={mockStore}>
        <CheckInBox {...props} />
      </TestingWrapperNavigation>
    );

    expect(component).toMatchSnapshot();
  });

  it("finds the check-in box", async () => {
    const component = render(
      <TestingWrapperNavigation store={mockStore}>
        <CheckInBox {...props} />
      </TestingWrapperNavigation>
    );

    const box = component.getByTestId("checkInBox");

    expect(box).toBeDefined();
  });
});
