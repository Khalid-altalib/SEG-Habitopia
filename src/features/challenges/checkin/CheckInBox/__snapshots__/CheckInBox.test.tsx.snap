// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckInBox renders the check-in box correctly 1`] = `
<RNCSafeAreaProvider
  onInsetsChange={[Function]}
  style={
    [
      {
        "flex": 1,
      },
      undefined,
    ]
  }
>
  <View
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <RNSScreenStack
      style={
        {
          "flex": 1,
        }
      }
    >
      <RNSScreen
        collapsable={false}
        forwardedRef={[Function]}
        gestureResponseDistance={
          {
            "bottom": -1,
            "end": -1,
            "start": -1,
            "top": -1,
          }
        }
        nativeBackButtonDismissalEnabled={false}
        onAppear={[Function]}
        onDisappear={[Function]}
        onDismissed={[Function]}
        onHeaderBackButtonClicked={[Function]}
        onNativeDismissCancelled={[Function]}
        onTransitionProgress={[Function]}
        onWillDisappear={[Function]}
        replaceAnimation="push"
        stackPresentation="push"
        style={
          {
            "bottom": 0,
            "left": 0,
            "position": "absolute",
            "right": 0,
            "top": 0,
          }
        }
        swipeDirection="horizontal"
      >
        <RNSScreenStackHeaderConfig
          backButtonInCustomView={false}
          backgroundColor="rgb(255, 255, 255)"
          color="rgb(0, 122, 255)"
          direction="ltr"
          disableBackButtonMenu={false}
          hidden={false}
          hideBackButton={false}
          largeTitleHideShadow={false}
          title="TestScreen"
          titleColor="rgb(28, 28, 30)"
          topInsetEnabled={false}
          translucent={false}
        />
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          style={
            {
              "flex": 1,
              "flexDirection": "column-reverse",
            }
          }
        >
          <View
            style={
              [
                {
                  "flex": 1,
                },
                {
                  "backgroundColor": "rgb(242, 242, 242)",
                },
                undefined,
              ]
            }
          >
            <View
              accessible={true}
              collapsable={false}
              focusable={true}
              onClick={[Function]}
              onResponderGrant={[Function]}
              onResponderMove={[Function]}
              onResponderRelease={[Function]}
              onResponderTerminate={[Function]}
              onResponderTerminationRequest={[Function]}
              onStartShouldSetResponder={[Function]}
              style={
                {
                  "marginRight": 25,
                  "opacity": 1,
                }
              }
              testID="checkInBox"
            >
              <View
                dataSet={{}}
                style={
                  [
                    {
                      "height": "100%",
                      "width": "100%",
                    },
                    {
                      "aspectRatio": 1,
                    },
                  ]
                }
              >
                <Image
                  accessibilityLabel="Fitness"
                  alt="Fitness"
                  dataSet={{}}
                  onError={[Function]}
                  source={
                    {
                      "uri": "https://picsum.photos/2000",
                    }
                  }
                  style={
                    [
                      {
                        "borderRadius": 8,
                        "height": "100%",
                        "maxWidth": "100%",
                        "position": "absolute",
                        "width": "100%",
                      },
                      {
                        "borderColor": "#00000011",
                        "borderWidth": 2,
                      },
                    ]
                  }
                />
                <Text
                  style={
                    [
                      {
                        "color": "white",
                      },
                      {
                        "fontFamily": undefined,
                        "fontSize": 19,
                      },
                      {
                        "padding": 12.25,
                        "textShadowColor": "black",
                        "textShadowRadius": 4,
                      },
                    ]
                  }
                >
                  Fitness
                </Text>
                <View
                  dataSet={{}}
                  style={
                    {
                      "bottom": 0,
                      "position": "absolute",
                    }
                  }
                >
                  <Text
                    style={
                      [
                        {
                          "color": "white",
                        },
                        {
                          "fontFamily": undefined,
                          "fontSize": 19,
                        },
                        {
                          "padding": 12.25,
                          "textShadowColor": "black",
                          "textShadowRadius": 4,
                        },
                      ]
                    }
                  >
                    NaNh left!
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </RNSScreen>
    </RNSScreenStack>
  </View>
</RNCSafeAreaProvider>
`;
